add result to after hook

Tomazh opened the pull request:
Hi, this is my first time making a PR here =)

I have an use case where I need to persist changes made on documents in a new collection and the after hook seemed like the perfect place to do that.
However this hook lacks the result from the `findOneAndUpdate` method, so I added it in there.
I also notice that there was not much going on with the variable `context` so I replaced it with the result

avaly commented on src/model.ts:
```diff
@@ -154,18 +154,17 @@ function wrap<A, R>(
   return async function modelWrapped(...args) {
     const { collectionName } = model.collection;
     const { after = [], before = [] } = model.options?.hooks || {};
-    const context = {};
```
Please do not remove the `context` variable. We rely on it at Plex to run some custom logic.

If you want to achieve your intended changes with this PR, just attach `result` in the `context` object.

avaly reviewed (changes requested):
This reminds me that we should document the hooks in detail.

Tomazh commented:
Ok I added back the context variable.
I also changed the position of the `result` parameter and made it optional so it wont be a breaking change when it gets deployed.

avaly reviewed (commented):
I still think that maybe the best way to implement this would be to attach `result` in the context object on line 166:

```ts
context.result = result;
```

avaly commented on src/hooks.ts:
```diff
@@ -7,7 +7,7 @@ export type Hook<A> = (
   args: A[],
   context: unknown,
-  error?: Error
+  error?: Error,
```
nit: please revert this change

avaly commented on src/__tests__/model.test.ts:
```diff
@@ -60,7 +60,7 @@ describe('model', () => {
   beforeEach(() => {
     doc = {
-      _id: new ObjectId(),
+      _id: new ObjectId('62b3876eb20d64c7bb577874'),
```
Why is this ID needed here?

Tomazh replied:
You're right, not needed anymore, I wasn't using the `result` variable to test the input parameters at first and was hard coding the expected input.

marafat commented:
I don't object to the idea but I feel `result` should be a first class argument on the hook just like `error`. This might mean that we will need a separate type for `HookAfter` and `HookBefore`, and it is going to be a breaking change, but I feel this is the right way to do it. 

`context` is not meant to have any typed members of the library. It is just a mean for the consumer of the library to pass some values in the before hooks and act on it in the after hook.

Tomazh commented:
Yeah I agree that it makes more sense to have a separate `result` type on each hook. We could avoid adding a breacking change by adding a overload of the hook function, what do you think?

Tomazh commented:
Yeah I agree that it makes more sense to add a separate type for hooks. We could also add `result` as an optional parameter right after `error`. Not pretty, I know, but it would avoid adding breaking changes:

src/hooks.ts
```
export type Hook<A> = (
  collectionName: string,
  methodName: string,
  args: A[],
  context: unknown,
  error?: Error,
  result?: unknown
) => Promise<void>;
```
src/model.ts
`await hook(collectionName, method.name, args, context, undefined, result );`

Tomazh commented:
Awesome, I'll wait for @avaly to give his opinion on this approach before making any changes

avaly commented:
I see your arguments for keeping the `result` separate from the `context` @marafat.

If we do add them as a separate argument to those hooks functions, and we're going to make this a breaking change anyway, let's transform the positional arguments for the hooks functions into a single object argument. This way any future additions to the input of the hooks, won't trigger a major version bump.

```ts
export type Hook<A> = (params: {
  args: A[];
  collectionName: string;
  context: unknown;
  error?: Error;
  methodName: string;
  result?: unknown;
}) => Promise<void>;
```

avaly reviewed (dismissed):
This looks great! Thanks for the contribution!

marafat commented on src/hooks.ts:
```diff
@@ -3,11 +3,14 @@ import { inspect } from 'util';
+    context: unknown;
+    error?: Error;
+    methodName: string;
+    result?: unknown;
```
I wonder if we can type this from the model 🤔

marafat reviewed (commented):
@avaly spends more time here than  I do, so if it looks good for him I am ok with it. I did notice some lint issues on unchanged files, worth taking a look at it. 
![CleanShot 2022-06-28 at 11 07 28](https://user-images.githubusercontent.com/3028854/176215060-ddc7ca22-e00d-42da-890f-fb42e49baee6.png)


Regarding my note on typing the `result`, I don't think it is a blocker but it is a nice follow up.

avaly commented:
> I did notice some lint issues on unchanged files, worth taking a look at it.

Nevermind those. They are old issues. We should suppress them at some point.

Tomazh commented:
Tks for all the help @avaly and @marafat!
I don't have write access to merge this PR, could one of you do the merge for me?
